### DISCUSSION TITLE
How can I improve my @next/mdx setup? (currently doing client side rendering when I know there's a better way)

### DISCUSSION THREAD

**Discussion by colinsteidtmann** [vercel/next.js#44314] [Help]
Current packages: `NextJS v.13, @next/mdx, @mdx-js/react`

I have a personal site and here's the folder structure, stored locally on GitHub within the project, that I like: 
|__content
|-- - author
| ------about
|--------- index.mdx
|--------- selfie.jpg
|-- - notes
| ------note 1
|---------  index.mdx
|---------  pic1.jpg
| ------note 2
|---------  index.mdx
|---------  pic2.jpg
|___src
|-- notes
| ------[slug].js
| ------index.js
|--  _app.js
|--   about.js
|--  index.js

I like to import my mdx content as a component and show it inside by js pages, like in `about.js` and `notes/[slug].js`. This is fine for static imports, but it's bad for dynamic imports such as in `[slug].js` where I use dynamic routing. Here's how I'm currently getting around it:

```js
// [slug].js
export default function Note({ metadata }) {
    const { title, date, tags, lastmod, type, slug } = metadata;
    return (
        <>
            <h1>{title}</h1>
            <MdxRenderer type={type} slug={slug} />
        </>
    );
}
```
```js
//components/MdxRenderer.js
// Client side rendering of the mdx file. TODO: Find a way to do this with pre-render
export default function MdxRenderer({ type, slug }) {
    const [mdxModule, setMdxModule] = useState();
    useEffect(() => {
        (async () => {
            const module = await import(`../content/${type}/${slug}/index.mdx`);
            setMdxModule(module);
        })();
    }, []);
    const Content = mdxModule ? mdxModule.default : Fragment;
    return <Content />;
}
```
See, the problem is that I'm client side rendering b/c I'm compiling the MDX file in `useEffect` when I know there should be a better way. The content is within the folder. I tried using NextJS `getStaticProps()` but dynamically importing the mdx module via `const module = await import(``../content/${type}/${slug}/index.mdx``);` returns a `MdxContent()` function to compile the content (I think). I can't run this function inside getStaticProps because it says something about "you can only use hooks inside function components." I also tried passing this as a prop to my component (by serializing/stringifying it inside `getStaticProps` and then de-serializing/rebuilding it inside the `Note` function component) but when I tried running it then I got an error saying I didn't have the webpack modules available that `MdxContent()` uses. This would've been pointless anyways b/c then I'd still be doing client side rendering and compiling the MDX content on the client side. 

TLDR; 
My Goal: compile the content as few times as compile and do it with pre-generation via getStaticProps or server-side
My Problem: I can't do this for dynamic paths to mdx content. 

--- Extra
I tried `mdx-bundler` but I don't like it because it uses esbuild when NextJS already has webpack/turbopack and they don't make it easy to use a provider like `@mdx-js/react` and using the package required adding lots of code bloat.

**Comment by GabenGar** [comment]
For "dynamic" paths you have to return the values of dynamic segments in `getStaticPaths()`. Then you'll be able just read and parse the file in `getStaticProps()` and return the metadata directly into the component.
